File: 90-subsets-ii/subsets-ii.cpp
class Solution {
public:
    void generateSubsets(int index,vector<int>& nums,vector<int>&subsets,vector<vector<int>>&ans)
    {
        ans.push_back(subsets);
        for(int i=index;i<nums.size();i++)
        {
            if(i>index && nums[i]==nums[i-1])continue;
            subsets.push_back(nums[i]);
            generateSubsets(i+1,nums,subsets,ans);
            subsets.pop_back();
        }
    }
    vector<vector<int>> subsetsWithDup(vector<int>& nums) 
    {
        sort(nums.begin(),nums.end());
        vector<int>subsets;
        vector<vector<int>>ans;

        generateSubsets(0,nums,subsets,ans);
        return ans;
    }
};

File: 90-subsets-ii/subsets-ii.java
class Solution {
    public void generateSubsets(int index,int n,List<List<Integer>> ans,List<Integer> subsets,int[] nums)
    {
        ans.add(new ArrayList(subsets));
        for(int i=index;i<n;i++)
        {
            if(index<i && nums[i]==nums[i-1])continue;
            subsets.add(nums[i]);
            generateSubsets(i+1,n,ans,subsets,nums);
            subsets.remove(subsets.size() - 1);
        }
        
    }
    public List<List<Integer>> subsetsWithDup(int[] nums) 
    {
        List<List<Integer>> ans= new ArrayList<>();    
        List<Integer> subsets=new ArrayList<>();
        Arrays.sort(nums);
        generateSubsets(0,nums.length,ans,subsets,nums);
        return ans;
    }
}